### PR TITLE
Closes DAS-1146: Sample weight metric calculations

### PR DESCRIPTION
**Pull request by ryan-deak-zefr**
# Summary 

This PR address that dask_ml (like sklearn) doesn't use the importance weights when calculating metrics when scoring folds in cross validation.  This affects all CV methods.

## Changes

This PR only addresses the use of `sample_weight` in the metric calculations in cross fold validation and doesn't address the issue of combining the metrics across folds, which is currently done by simple averaging and not by weighted averaging using the sum of the sample weights for a fold as the weight.

## Tests

`test_sample_weight_in_metrics` in `tests/model_selection/dask_searchcv/test_model_selection.py`.

This test is written to pass given the current way of combining metrics across folds (which is likely not the best / correct way to do it).

### CODE REVIEW

File: dask_ml/model_selection/methods.py [hunk 1]
```diff
@@ -276,19 +276,135 @@ def fit_transform(
...
+        train_sample_weight = None
+        test_sample_weight = None
+    else:
+        # "0" is the train split, "1" is the test split.
+        train_sample_weight = safe_indexing(sample_weight, cv.splits[n][0])
+        test_sample_weight = safe_indexing(sample_weight, cv.splits[n][1])
+    return train_sample_weight, test_sample_weight
+
+
+def _apply_scorer(estimator, X, y, scorer, sample_weight):
```
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/methods.py:323]: Same as in the sklearn PR for DAS-1145.

File: dask_ml/model_selection/methods.py [hunk 2]
```diff
@@ -276,19 +276,135 @@ def fit_transform(
...
-def score(est_and_time, X_test, y_test, X_train, y_train, scorer, error_score):
+def score(
+    est_and_time,
+    X_test,
+    y_test,
+    X_train,
+    y_train,
+    scorer,
+    error_score,
+    test_sample_weight=None,
```
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/methods.py:409]: Ordering is consistent with the `test` then `train` parameter order established above.

File: dask_ml/model_selection/_search.py [hunk 3]
```diff
@@ -275,6 +276,17 @@ def do_fit_and_score(
     scorer,
     return_train_score,
 ):
+    if "sample_weight" in fit_params:
```
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/_search.py:286]: At the top since this is used in both branches.

File: dask_ml/model_selection/_search.py [hunk 4]
```diff
@@ -275,6 +276,17 @@ def do_fit_and_score(
...
+        # This will likely get all sample weights but we might want to
+        # whittle this down since it'll ultimately be used to get the
+        # test sample weights.
+        #
+        # Each value in the fit_params dict is a 2-tuple where the
+        # data representation is in the second dimension (dim 1).
+        sample_weight = fit_params["sample_weight"][1]
```
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/_search.py:293]: Not sure about getting this from some cache or something.

File: dask_ml/model_selection/methods.py [hunk 5]
```diff
@@ -314,16 +430,32 @@ def fit_and_score(
     params=None,
     fit_params=None,
     return_train_score=True,
+    sample_weight=None,
```
Comment (ryan-deak-zefr) [inline dask_ml/model_selection/methods.py:441]: Place at the end and make it `None` by default to make it "backward compatible".

File: tests/model_selection/dask_searchcv/test_model_selection.py [hunk 6]
```diff
@@ -891,3 +898,56 @@ def test_mock_with_fit_param_raises():
 
     with pytest.raises(ValueError):
         clf.fit(X, y)
+
+
+@pytest.mark.parametrize(
```
Comment (ryan-deak-zefr) [inline tests/model_selection/dask_searchcv/test_model_selection.py]: This test kind of subsumes the test in the sklearn PR because it tests the situation with one fold metric calculation and multiple fold metric calculations that need to be combined.  It does this by testing at the `dcv.GridSearchCV` level rather than at the level of functions in `dask_ml/model_selection/methods.py`.

**Review by ryan-deak-zefr — COMMENTED** [review]
This works to fix some of the issues but there are presumably much deeper methodological issues that might make this PR moot.

File: tests/model_selection/dask_searchcv/test_model_selection.py [hunk 7]
```diff
@@ -891,3 +898,56 @@ def test_mock_with_fit_param_raises():
...
+            [("train", "test"), ("test", "train")],
+            (1 / (1 + 99) + 100 / (100 + 200)) / 2,
+        ),
+        (
+            [("test", "train"), ("train", "test")],
+            (100 / (100 + 200) + 1 / (1 + 99)) / 2,
+        ),
+    ],
+)
+def test_sample_weight_in_metrics(cv_ind, exp_acc):
```
Comment (ryan-deak-zefr) [inline tests/model_selection/dask_searchcv/test_model_selection.py]: **NOTE**: If this test were used in the current master of *dask_ml*, it would fail because the metric returned by all tests is `0.5`.

`0.5` **is *definitely* incorrect** as it disregards `sample_weight` entirely.

**Review by zexuan-zhou — APPROVED** [review]
Looks all good to me

File: dask_ml/model_selection/methods.py [hunk 8]
```diff
@@ -276,19 +276,135 @@ def fit_transform(
...
-def _score(est, X, y, scorer):
+def _get_fold_sample_weights(sample_weight, cv, n):
+    """Get the training and test sample weights for the ``n`` th fold using
+    ``cv``.
+
+    Parameters
+    ----------
+    sample_weight : array-like
+        sample weights.  Should contain all sample weights needed for
+        training and testing. May be None.
```
Comment (jon-morra-zefr) [inline dask_ml/model_selection/methods.py:287]: Why is it sensical for `sample_weight` to be `None` in a function called `_get_fold_sample_weights`?

File: dask_ml/model_selection/methods.py [hunk 9]
```diff
@@ -276,19 +276,135 @@ def fit_transform(
...
+    test_sample_weight : array-like
+    """
+
+    # NOTE: train split is unnecessary if X_train is None b/c of the check in
+    #       ``score``.  That's an optimization that can be addressed later.
+    if sample_weight is None:
+        train_sample_weight = None
+        test_sample_weight = None
+    else:
+        # "0" is the train split, "1" is the test split.
```
Comment (jon-morra-zefr) [inline dask_ml/model_selection/methods.py:309]: Is this a convention somewhere else?

File: tests/model_selection/dask_searchcv/test_model_selection.py [hunk 10]
```diff
@@ -5,6 +5,11 @@
 from itertools import product
 from multiprocessing import cpu_count
 
+# sklearn.metrics.make_scorer
```
Comment (jon-morra-zefr) [inline tests/model_selection/dask_searchcv/test_model_selection.py]: What is the point of these comments?

**Review by vecchp — APPROVED** [review]
+1 Let' just make sure we issue the sklearn pr first before this goes out.

**Comment by ryan-deak-zefr** [comment]
This PR has the same tests as the scikit-learn PR: https://github.com/ZEFR-INC/scikit-learn/pull/2  So it should be working very similarly.
